modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java

```diff
@@ -287,4 +365,115 @@ void countVersionConflict() {
     void countRetry() {
         retries.incrementAndGet();
     }
-}
\ No newline at end of file
+
+    float getRequestsPerSecond() {
+        return requestsPerSecond;
+    }
+
+    /**
+     * Schedule prepareBulkRequestRunnable to run after some delay. This is where throttling plugs into reindexing so the request can be
+     * rescheduled over and over again.
+     */
+    void delayPrepareBulkRequest(ThreadPool threadPool, TimeValue delay, final AbstractRunnable prepareBulkRequestRunnable) {
+        // Synchronize so we are less likely to schedule the same request twice.
+        synchronized (delayedPrepareBulkRequestReference) {
+            AbstractRunnable oneTime = new AbstractRunnable() {
+                private final AtomicBoolean hasRun = new AtomicBoolean(false);
+
+                @Override
+                protected void doRun() throws Exception {
+                    // Paranoia to prevent furiously rethrottling from running the command multiple times. Without this we totally can.
+                    if (hasRun.compareAndSet(false, true)) {
+                        prepareBulkRequestRunnable.run();
+                    }
+                }
+
+                @Override
+                public void onFailure(Throwable t) {
+                    prepareBulkRequestRunnable.onFailure(t);
+                }
+            };
+            delayedPrepareBulkRequestReference.set(new DelayedPrepareBulkRequest(threadPool, getRequestsPerSecond(), delay, oneTime));
+        }
+    }
+
+    private void setRequestsPerSecond(float requestsPerSecond) {
+        if (requestsPerSecond == -1) {
+            requestsPerSecond = 0;
+        }
+        this.requestsPerSecond = requestsPerSecond;
+    }
+
+    void rethrottle(float newRequestsPerSecond) {
+        synchronized (delayedPrepareBulkRequestReference) {
+            setRequestsPerSecond(newRequestsPerSecond);
+
+            DelayedPrepareBulkRequest delayedPrepareBulkRequest = this.delayedPrepareBulkRequestReference.get();
+            if (delayedPrepareBulkRequest == null) {
+                // No request has been queued yet so nothing to reschedule.
+                return;
+            }
+
+            this.delayedPrepareBulkRequestReference.set(delayedPrepareBulkRequest.rethrottle(newRequestsPerSecond));
```

Comment:
I believe that to be safe because we're doing it within a synchronized block. And I used a synchronized block because this is low contention and I wanted there to be less of a chance that we'd queue two copies of the request.